analysis/import/SourceCodeParser/src/main/kotlin/de/maibornwolff/codecharta/importer/sourcecodeparser/visitors/MaxNestingLevelVisitor.kt

```diff
@@ -0,0 +1,116 @@
+package de.maibornwolff.codecharta.importer.sourcecodeparser.visitors
+
+import org.sonar.java.ast.visitors.ComplexityVisitor
+import org.sonar.plugins.java.api.tree.*
+
+class MaxNestingLevelVisitor: ComplexityVisitor() {
+    private var maxNestingLevel = 0
+    private var forwardNesting = 0
+
+    fun getMaxNestingLevel(tree: Tree): Int {
+        super.getNodes(tree)
+
+        return maxNestingLevel
+    }
+
+    private fun checkForChildrenAndUpdateMax() {
+        if (maxNestingLevel < forwardNesting) {
+            maxNestingLevel = forwardNesting
+        }
+    }
+
+    override fun visitMethod(tree: MethodTree) {
+        forwardNesting = 0
+        super.visitMethod(tree)
+    }
+
+    override fun visitIfStatement(tree: IfStatementTree) {
+        forwardNesting++
+
+        checkForChildrenAndUpdateMax()
+
+        super.visitIfStatement(tree)
+
+        forwardNesting--
+    }
+
+    override fun visitDoWhileStatement(tree: DoWhileStatementTree) {
+        forwardNesting++
+
+        checkForChildrenAndUpdateMax()
+
+        super.visitDoWhileStatement(tree)
+
+        forwardNesting--
+    }
+
+    override fun visitLambdaExpression(lambdaExpressionTree: LambdaExpressionTree) {
```

Comment:
@Richargh does a lambda-function increase the nesting level by one?

Reply:
I couldn't find a formal definition, but I would say yes.